File: Arrays_and_Hashing/C++/217_Contains_duplicate.cpp
/*
* TIME COMPLEXITY - O(nlogn)
* RUNTIME: 91 ms - Beats 99.47% of submissions
* MEMORY - 57.1 MB - Beats 100% of submissions
*/

#include <vector>
#include <algorithm>
#include <iostream>

/*
* Thoughtprocess:
* If the Array is empty or has only one element, return false
* Sort the array
* Iterate through the array and check if the current element is equal to the next element
*/
class Solution {
public:
    bool containsDuplicate(std::vector<int>& nums) {
        if (nums.size() <= 1)
            return false;
        sort(nums.begin(), nums.end());
        int check =nums[1];

        for (int num : nums)
        {
            if (check == num)
                return true;
            else
                check = num;
        }
        return false;
    }
};

File: Arrays_and_Hashing/Python/217_contains_duplicate.py


# My thougtprocess:
# 1. Sort the array
# 2. Iterate through the array and check if the current element is equal to the previous element
# 3. If yes, return True

# Time complexity: O(nlogn)
# At fist I thought about using two loops, but that would have been O(n^2) time complexity
# Sorting seemed like a good idea, even though it could be probably optimized further. 

# RUNTIME: 507 ms, faster than 13.18% of python3 Submissions. 
# MEMORY USAGE: 28.2 MB, less than 95.25% of python3 Submissions.


from __future__ import annotations
from ast import *

class Solution:
    def containsDuplicate(self, nums: List[int]) -> bool:
        nums.sort()
        for index, num in enumerate(nums):
            if index == 0:
                continue
            else:
                if nums[index-1] == num:
                    return True